ortb/clone_test.go

```diff
@@ -0,0 +1,693 @@
+package ortb
+
+import (
+	"encoding/json"
+	"reflect"
+	"testing"
+
+	"github.com/prebid/openrtb/v17/adcom1"
+	"github.com/prebid/openrtb/v17/openrtb2"
+	"github.com/prebid/prebid-server/util/ptrutil"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCloneApp(t *testing.T) {
+	t.Run("nil", func(t *testing.T) {
+		result := CloneApp(nil)
+		assert.Nil(t, result)
+	})
+
+	t.Run("populated", func(t *testing.T) {
+		given := &openrtb2.App{
+			ID:                     "anyID",
+			Name:                   "anyName",
+			Bundle:                 "anyBundle",
+			Domain:                 "anyDomain",
+			StoreURL:               "anyStoreURL",
+			CatTax:                 adcom1.CatTaxIABContent10,
+			Cat:                    []string{"cat1"},
+			SectionCat:             []string{"sectionCat1"},
+			PageCat:                []string{"pageCat1"},
+			Ver:                    "anyVer",
+			PrivacyPolicy:          1,
+			Paid:                   2,
+			Publisher:              &openrtb2.Publisher{ID: "anyPublisher", Ext: json.RawMessage(`{"publisher":1}`)},
+			Content:                &openrtb2.Content{ID: "anyContent", Ext: json.RawMessage(`{"content":1}`)},
+			Keywords:               "anyKeywords",
+			KwArray:                []string{"key1"},
+			InventoryPartnerDomain: "anyInventoryPartnerDomain",
+			Ext:                    json.RawMessage(`{"anyField":1}`),
+		}
+		result := CloneApp(given)
+		assert.Equal(t, given, result, "equality")
+		assert.NotSame(t, given, result, "pointer")
+		assert.NotSame(t, given.Cat, result.Cat, "cat")
+		assert.NotSame(t, given.SectionCat, result.SectionCat, "sectioncat")
+		assert.NotSame(t, given.PageCat, result.PageCat, "pagecat")
+		assert.NotSame(t, given.Publisher, result.Publisher, "publisher")
+		assert.NotSame(t, given.Publisher.Ext, result.Publisher.Ext, "publisher-ext")
+		assert.NotSame(t, given.Content, result.Content, "content")
+		assert.NotSame(t, given.Content.Ext, result.Content.Ext, "content-ext")
+		assert.NotSame(t, given.KwArray, result.KwArray, "kwarray")
+		assert.NotSame(t, given.Ext, result.Ext, "ext")
+	})
+
+	t.Run("assumptions", func(t *testing.T) {
+		assert.ElementsMatch(t, discoverPointerFields(reflect.TypeOf(openrtb2.App{})),
+			[]string{
+				"Cat",
+				"SectionCat",
+				"PageCat",
+				"Publisher",
+				"Content",
+				"KwArray",
+				"Ext",
+			})
+	})
```

Comment:
Assumptions test protect against new pointer fields which need to be specifically accounted for in the clone methods. Expected only to break when updating the openrtb library.